core/src/main/java/com/datastax/oss/driver/internal/core/cql/CqlPrepareAsyncProcessor.java

```diff
@@ -21,36 +21,44 @@
 import com.datastax.oss.driver.api.core.type.reflect.GenericType;
 import com.datastax.oss.driver.internal.core.context.InternalDriverContext;
 import com.datastax.oss.driver.internal.core.session.DefaultSession;
-import com.datastax.oss.driver.internal.core.session.RequestHandler;
 import com.datastax.oss.driver.internal.core.session.RequestProcessor;
-import java.nio.ByteBuffer;
+import com.datastax.oss.driver.internal.core.util.concurrent.CompletableFutures;
+import com.datastax.oss.driver.shaded.guava.common.cache.Cache;
+import com.datastax.oss.driver.shaded.guava.common.cache.CacheBuilder;
+import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CompletionStage;
-import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutionException;
 import net.jcip.annotations.ThreadSafe;
 
 @ThreadSafe
 public class CqlPrepareAsyncProcessor
     implements RequestProcessor<PrepareRequest, CompletionStage<PreparedStatement>> {
 
-  private final ConcurrentMap<ByteBuffer, DefaultPreparedStatement> preparedStatementsCache;
-
-  public CqlPrepareAsyncProcessor(
-      ConcurrentMap<ByteBuffer, DefaultPreparedStatement> preparedStatementsCache) {
-    this.preparedStatementsCache = preparedStatementsCache;
-  }
+  private final Cache<PrepareRequest, CompletableFuture<PreparedStatement>> cache =
+      CacheBuilder.newBuilder().weakValues().build();
 
   @Override
   public boolean canProcess(Request request, GenericType<?> resultType) {
     return request instanceof PrepareRequest && resultType.equals(PrepareRequest.ASYNC);
   }
 
   @Override
-  public RequestHandler<PrepareRequest, CompletionStage<PreparedStatement>> newHandler(
+  public CompletionStage<PreparedStatement> process(
       PrepareRequest request,
       DefaultSession session,
       InternalDriverContext context,
       String sessionLogPrefix) {
-    return new CqlPrepareAsyncHandler(
-        request, preparedStatementsCache, session, context, sessionLogPrefix);
+
+    try {
+      return cache.get(
+          request,
```

Comment:
So this is using `PrepareRequest` as the entry key. I  see that you added equals/hashCode implementations where appropriate, but a quick look into the implementations shows that `DefaultPrepareRequest.equals()` will call `SimpleStatement.equals()`, but `DefaultSimpleStatement.equals()` has a very complex equals method that compares much more than the query string. This could lead to a situation where two non-equal `SimpleStatement` instances sharing the same query string would actually generate two cache entries. Is this a problem?

Reply:
It's not a bug, it's a feature 😉 
It does generate separate cache entries, but I see that as an improvement. If you prepare with different execution parameters (e.g. same query but different CL), each prepared statement generates bound statements that honor those.

See also `PreparedStatementIT.should_create_separate_instances_for_different_statement_parameters()` and my comment above.

Reply:
Ah nice, indeed 👍